adam/flask_backend/app.py

```diff
@@ -135,7 +144,168 @@ def get_scene() -> Any:
     }
 
 
+@app.route("/api/get_image", methods=["GET"])
+def get_image() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    learner = request.args.get("learner", default="")
+    training_curriculum = request.args.get("training_curriculum", default="")
+    testing_curriculum = request.args.get("testing_curriculum", default="")
+    scene_number = int(request.args.get("scene_number", "")) - 1
+    situation_dir = f"situation_{scene_number}"
+    image_file = request.args.get("image_file", default="")
+    if (
+        not learner
+        or not training_curriculum
+        or not testing_curriculum
+        or not situation_dir
+    ):
+        abort(HTTPStatus.BAD_REQUEST)
+
+    if any(
+        "/" in arg
+        for arg in (
+            learner,
+            training_curriculum,
+            testing_curriculum,
+            situation_dir,
+            image_file,
+        )
+    ):
+        abort(HTTPStatus.BAD_REQUEST)
+    image_path = (
+        LEARNERS_DIR
+        / learner
+        / EXPERIMENTS_DIR_NAME
+        / training_curriculum
+        / EXPERIMENTS_TESTING_DIR_NAME
+        / testing_curriculum
+        / situation_dir
+        / image_file
+    ).resolve()
+    if not is_relative_to(image_path, DATA_DIR):
+        return {"message": "Directory out of range"}, HTTPStatus.BAD_REQUEST
+
+    if (
+        learner not in app.config["LEARNER_DIR_CONTENTS"]
+        or training_curriculum not in app.config["LEARNER_DIR_CONTENTS"][learner]
+        or testing_curriculum
+        not in app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum]
+        or situation_dir
+        not in app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum][
+            testing_curriculum
+        ]
+        or image_file
+        not in app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum][
+            testing_curriculum
+        ][situation_dir]
```

Comment:
Hm, I just realized technically this part is probably a bit too restrictive. In the absence of the frontend, any requests for images generated after the server starts would be rejected. But I'm okay with that because in practice we would only use this backend together with the frontend, and if we ended up needing this to work in another context it's a pretty easy fix.

Reply:
Right, currently the `LEARNER_DIR_CONTENTS` only updates if `poll_queues` is invoked but if we want it to function independently of that function we can either drop the check or build the full file path and check if the file exists.